hedera-node/docs/design/services/token-service/airdrops/airdrop-pending-state.md

```diff
@@ -0,0 +1,126 @@
+3. Describe how this new state would be serialized and kept after restarts
+4. Define Readable and Writable stores for accessing and manipulating the pending airdrops state
+
+## Architecture
+
+The newly defined entity state won’t be able to persist anything in the state directly. If we want to commit something from the pending state to the persisted state, we would have to delegate a given pending airdrop by creating a synthetic crypto transfer transaction, send it to a suitable handler and delete it from the pending state. The writable nature of the pending state would be limited only to adding new entries for a pending airdrop to be transferred or to remove already added ones.
+
+The `WritableStore` of the airdrops, should contain a handler method and if a given pending airdrop is claimed, it could be delegated to the `CryptoTransferHandler` and executed.
+
+Details for `TokenClaimAirdrop` and `TokenCancelAirdrop` would be covered in their own design, docs but basically the `claim` operation would take a pending airdrop and submit it to the delegate function for actual handling and execution. The `cancel` operation would clear an added pending airdrop from the state and it won’t be executed at all.
+
+### Structure
+
+We should define the structure of pending changes that would be kept in the `InitialModServiceTokenSchema` state, so we would reuse the token’s serialized state and add pending airdrops records in it. We could use the type defined in the protobuf records for the airdrop information itself. We should use the protobuf types since they are used for state serialization. So, the structure could look like this:
+
+`StateDefinition.*onDisk*(*AIRDROPS_KEY*, PendingAirdropId.PROTOBUF, PendingAirdropValue.PROTOBUF, MAX__AIRDROPS);`
+
+We can have a case, where we have multiple airdrops for the same fungible token. In this case the  `PendingAirdropValue` would be updated with the new aggregated value.
+
+### Serialization
+
+In order for the pending state to be kept during node upgrades or restarts we should write it on disk with a serialization mechanism, as we do for other context or state related information. This could be done in the `InitialModServiceTokenSchema` since the pending airdrop state would be written in the already existing token state.
+
+We should define a new entry in `StateChildIndices` called `AIRDROPS`. Based on it, we would fetch a `VirtualMap` entry for the airdrops. It would include an `AirdropKey`, which would be a `PendingAirdropId` representation as a key and an `EntityNum` , which would be a `PendingAirdropValue` representation as a value. The `AirdropKey` should implement the `VirtualKey` interface, so that the key has a very strong hash algorithm to optimize performance. `EntityNum` is an already existing numeric wrapper for long values that extend the `VirtualValue` interface and provide better efficiency for usage in `VirtualMaps`.
```

Comment:
As discussed today, we don't need this section. And this doc should be updated design for traversal of all pending airdrops sent by an account (using the LinkedList pattern) during expiration or deletion

Reply:
I adapted the doc following your comments.